Bulk processor concurrent requests

jakelandis opened the pull request:
`org.elasticsearch.action.bulk.BulkProcessor` is a threadsafe class that
allows for simple semantics to deal with sending bulk requests. Once a
bulk reaches it's pre-defined size, documents, or flush interval it will
execute sending the bulk. One configurable option is the number of concurrent
outstanding bulk requests. That concurrency is implemented in
`org.elasticsearch.action.bulk.BulkRequestHandler` via a semaphore. However,
the only code that currently calls into this code is blocked by `synchronized`
methods. This results in the in-ability for the BulkProcessor to behave concurrently
despite supporting configurable amounts of concurrent requests.

This change removes the `synchronized` method in favor an explicit
lock around the non-thread safe parts of the method. The call into
`org.elasticsearch.action.bulk.BulkRequestHandler` is no longer blocking, which
allows `org.elasticsearch.action.bulk.BulkRequestHandler` to handle it's own concurrency.



---

Currently the only production code that uses this class is Watcher
and it is defaulted to 0 concurrency. This change allows Watcher to optionally 
increase the concurrency.

I have run this test a few thousand times and can not get it to error. Also, if I remove the lock locally the test will fail pretty quickly.  

This is tangentially related to https://github.com/elastic/elasticsearch/pull/41418 since it is the same synchronized method that can cause issues. However, this change without that fix only moves where the blocking happens (from synchronized block to semaphore block). 

Adding HLRC and Watcher since outside of tests, they are the largest consumers.

jakelandis committed:
Allow bulkProcessor to behave concurrently

  `org.elasticsearch.action.bulk.BulkProcessor` is a threadsafe class that
  allows for simple semantics to deal with sending bulk requests. Once a
  bulk reaches it's pre-defined size, documents, or flush interval it will
  execute sending the bulk. One configurable option is the number of concurrent
  outstanding bulk requests. That concurrency is implemented in
  `org.elasticsearch.action.bulk.BulkRequestHandler` via a semaphore. However,
  the only code that currently calls into this code is blocked by `sychrnoized`
  methods. This results in the in-ability for the BulkProcessor to behave concurrently
  despite supporting configurable amounts of concurrent requests.

  This change removes the `synchronized` method in favor an explicit
  lock around the non-thread safe parts of the method. The call into
  `org.elasticsearch.action.bulk.BulkRequestHandler` is no longer blocking, which
  allows `org.elasticsearch.action.bulk.BulkRequestHandler` to handle it's own concurrency.

  Note - only the primary add method has been updated since this is the hot path.
  Currently only Watcher uses this class and the concurrency is set to 0 by default.
  This change allows Watcher to optionally increase the concurrency.

jakelandis committed:
fix rouge formatting

elasticmachine commented:
Pinging @elastic/es-core-features

jakelandis commented:
Hold off reviewing...I think I need to wrap the other instance swapping inside the same lock.

jakelandis committed:
replace synchronized methods with same lock

jakelandis committed:
allow other add to be non-synchronous and add more tests

jakelandis committed:
add missing ensure open

jakelandis committed:
fix checkstyle and update comments

jakelandis committed:
add some delay to simulate work and fix forbiden api check

jakelandis committed:
ensure test is under ~10 seconds

jakelandis committed:
tweak test

jakelandis committed:
fix divide by zero

jakelandis committed:
fix checkstyle

jakelandis committed:
fix forbiddenAPI

jakelandis committed:
bump up future.get max wait time

jakelandis commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

jakelandis commented:
@martijnvg @hub-cap - this is ready for review. I have re-run the tests a few hundred more times with the code as-is and can't get it to fail. To see this change in action you need to add the following println in `BulkRequestHandler` and run the concurrency tests. 

```
   public void execute(BulkRequest bulkRequest, long executionId) {
        Runnable toRelease = () -> {};
        boolean bulkRequestSetupSuccessful = false;
        try {
            listener.beforeBulk(executionId, bulkRequest);
            semaphore.acquire();
            System.out.println("permits:" + semaphore.availablePermits() + " of " + concurrentRequests + " queue length: " + semaphore.getQueueLength());

```
You will notice that before this change, it at most ever acquired 1 permit (due to the upstream synchronous method). With this change it can consume all the permits and add to the queue.

jakelandis committed:
Merge branch 'master' into bulk_processor_concurrent_requests

hub-cap commented on server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java:
```diff
@@ -264,21 +267,26 @@ public void close() {
-        closed = true;
+    public boolean awaitClose(long timeout, TimeUnit unit) throws InterruptedException {
+        try {
+            lock.lock();
```
I did some quick searching here because Im just not too familiar w/ this stuff, and it looks like most of the time the `lock.lock()` is outside of the try, which i assume is to throw the `InterruptedException`, rather than to execute the `lock.unlock()` before throwing. Should the `lock.lock()` be in the `try` ?

jaymode replied:
The `lock.lock()` should be outside of the try. If the lock is not obtained then calling unlock will throw an `IllegalMonitorStateException` since the thread does not hold the lock.

hub-cap replied:
ty for the explain!!

jakelandis replied:
thanks @baz and @jaymode - fixed in https://github.com/elastic/elasticsearch/pull/41451/commits/a9e0f3bd8db3b7aa79ba2166895057c198475f56

hub-cap reviewed (approved):
LGTM thank you for the detailed commit msg and note about the testing. I had one question but I will defer to those who are more well versed than I in the art of locks in java!

jaymode commented on server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java:
```diff
@@ -219,12 +221,13 @@ private static Scheduler buildScheduler(ScheduledThreadPoolExecutor scheduledThr
     private final AtomicLong executionIdGen = new AtomicLong();
 
-    private BulkRequest bulkRequest;
+    private volatile BulkRequest bulkRequest;
```
volatile should not be necessary here given that all access is done under lock

jakelandis replied:
fixed in https://github.com/elastic/elasticsearch/pull/41451/commits/a9e0f3bd8db3b7aa79ba2166895057c198475f56

jaymode commented on server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java:
```diff
@@ -264,21 +267,26 @@ public void close() {
+            } finally {
+                onClose.run();
+            }
+        }finally {
```
```suggestion
        } finally {
```

jakelandis replied:
fixed in https://github.com/elastic/elasticsearch/pull/41451/commits/a9e0f3bd8db3b7aa79ba2166895057c198475f56

jaymode commented on server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java:
```diff
@@ -388,25 +429,32 @@ private boolean isOverTheLimit() {
+            if (bulkRequest.numberOfActions() > 0) {
+                execute();
+            }
+        }finally {
```
```suggestion
        } finally {
```

jakelandis replied:
fixed in https://github.com/elastic/elasticsearch/pull/41451/commits/a9e0f3bd8db3b7aa79ba2166895057c198475f56

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,174 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        //find some randoms that allow this test to take under ~ 10 seconds
+        while (estimatedTimeForTest > 10_000) {
+            if (iterations++ > 1_000) {
+                fail("failed to find random values that allows test to run quickly"); //extremely unlikely
```
instead of failing the test, use an assumeTrue/assumeFalse?

jakelandis replied:
good suggestion! done - https://github.com/elastic/elasticsearch/pull/41451/commits/a1e456c75a97ee28ff79695afc96c8503d34bf63

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,174 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        for (final AtomicInteger i = new AtomicInteger(0); i.getAndIncrement() < maxDocuments;) {
+            //alternate between ways to add to the bulk processor
+            futures.add(executorService.submit(() -> {
+                if(i.get() % 2 == 0) {
```
I am not sure this is the best way to test this. Ideally you want all of your threads ready to go prior to starting work.  Usually we use a cyclic barrier or countdown latch so that we know all threads are ready prior to execution of the concurrent tasks. An example use of a cyclic barrier would be https://github.com/elastic/elasticsearch/blob/0531987a0466fd7e4af367c76cf6b8911b4df51c/server/src/test/java/org/elasticsearch/common/cache/CacheTests.java#L843-L878.

jakelandis replied:
@jaymode - thanks for the pointer. I added countdown latch to ensure that all the tasks are submitted to the executor before doing any work. Let me know if this missed the mark. - https://github.com/elastic/elasticsearch/pull/41451/commits/a1e456c75a97ee28ff79695afc96c8503d34bf63

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,174 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        for (Future f : futures) {
+            try {
+                f.get(10, TimeUnit.SECONDS);
+            }catch (Exception e){
```
```suggestion
            } catch (Exception e){
```

jakelandis replied:
fixed in https://github.com/elastic/elasticsearch/pull/41451/commits/a9e0f3bd8db3b7aa79ba2166895057c198475f56

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,174 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+
+        for (Future f : futures) {
+            try {
+                f.get(10, TimeUnit.SECONDS);
```
typically we avoid things like since timing could get messed up by a slow machine or GC pause. We can just use `Future#get`

jakelandis replied:
++ fixed - https://github.com/elastic/elasticsearch/pull/41451/commits/a1e456c75a97ee28ff79695afc96c8503d34bf63

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,174 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        for (Future f : futures) {
+            try {
+                f.get(10, TimeUnit.SECONDS);
+            }catch (Exception e){
```
```suggestion
            } catch (Exception e){
```

jakelandis replied:
fixed in https://github.com/elastic/elasticsearch/pull/41451/commits/a9e0f3bd8db3b7aa79ba2166895057c198475f56

jaymode reviewed (changes requested):
I left some comments about lock usage and the testing

jakelandis committed:
move lock outside of try block and fix formatting

jakelandis committed:
improve tests based on feedback

jakelandis commented:
@jaymode @hub-cap - thanks for your reviews! I believe all comments have been addressed.

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                Thread.sleep(simulateWorkTimeInMillis); //simulate work
+            } catch (InterruptedException e) {
+                failureCount.getAndIncrement();
+                logger.error("interrupted while sleeping. There is likely something wrong with this test!", e);
```
```suggestion
                logger.error("interrupted while sleeping. There is likely something wrong with this test!", e);
                exceptionRef.set(ExceptionsHelper.useOrSuppress(exceptionRef.get(), e);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+            futures.add(executorService.submit(() -> {
+                try {
+                    //don't start any work until all tasks are submitted
+                    startGate.await();
```
```suggestion
                    startGate.countDown();
                    startGate.await();
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                }
+            }));
+        }
+        startGate.countDown();
```
```suggestion
        startGate.countDown();
        startGate.await();
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        for (final AtomicInteger i = new AtomicInteger(0); i.getAndIncrement() < maxDocuments;) {
+            //alternate between ways to add to the bulk processor
+            futures.add(executorService.submit(() -> {
+                if(i.get() % 2 == 0) {
```
```suggestion
                //don't start any work until all tasks are submitted
                startGate.countDown();
                startGate.await();
                if (randomBoolean()) {
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                failureCount.getAndIncrement();
+                logger.error("interrupted while sleeping. There is likely something wrong with this test!", e);
+            }
+            listener.onResponse(bulkResponse);
```
I wonder if we should make the consumer here and the next test consistent. This one calls onResponse regardless of if an InterruptedException is thrown. The next test only calls if the sleep was successful.

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+            futures.add(executorService.submit(() -> {
+                if(i.get() % 2 == 0) {
+                    bulkProcessor.add(indexRequest);
+                }else{
```
```suggestion
                } else {
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                listener.onResponse(bulkResponse);
+            } catch (InterruptedException e) {
+                failureCount.getAndIncrement();
+                logger.error("interrupted while sleeping. There is likely something wrong with this test!", e);
```
```suggestion
                logger.error("interrupted while sleeping. There is likely something wrong with this test!", e);
                exceptionRef.set(ExceptionHelper.useOrSuppress(exceptionRef.get(), e);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        final int maxBatchSize = Integer.MAX_VALUE; //don't flush based on size
+        final int concurrentBulkRequests = randomIntBetween(0, 20);
+        final int simulateWorkTimeInMillis = 5;
+        BulkResponse bulkResponse = new BulkResponse(new BulkItemResponse[]{new BulkItemResponse()}, 0);
```
this is just my preference, feel free to ignore:

```suggestion
        BulkResponse bulkResponse = new BulkResponse(new BulkItemResponse[] { new BulkItemResponse() }, 0);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                    try {
+                        bulkProcessor.add(bytesReference, null, null, XContentType.JSON);
+                    } catch (Exception e) {
+                        throw new RuntimeException(e);
```
```suggestion
                        throw ExceptionsHelper.convertToRuntime(e);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+
+        for (Future f : futures) {
+            try {
+                f.get(10, TimeUnit.SECONDS);
```
```suggestion
                f.get();
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        for (Future f : futures) {
+            try {
+                f.get(10, TimeUnit.SECONDS);
+            } catch (Exception e){
```
```suggestion
            } catch (Exception e) {
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
     }
 
+    public void testConcurrentExecutions() throws Exception {
+        final AtomicBoolean called = new AtomicBoolean(false);
```
```suggestion
        final AtomicBoolean called = new AtomicBoolean(false);
        final AtomicReference<Exception> exceptionRef = new AtomicReference<>();
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                        bulkProcessor.add(bytesReference, null, null, XContentType.JSON);
+                    }
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
```
```suggestion
                    throw ExceptionHelper.convertToRuntime(e);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                f.get();
+            } catch (Exception e){
+                failureCount.incrementAndGet();
+                logger.error("failure while getting future", e);
```
```suggestion
                logger.error("failure while getting future", e);
                exceptionRef.set(ExceptionsHelper.useOrSuppress(exceptionRef.get(), e);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        for (Future f : futures) {
+            try {
+                f.get();
+            } catch (Exception e){
```
```suggestion
            } catch (Exception e) {
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        executorService.awaitTermination(10, TimeUnit.SECONDS);
+
+        if (failureCount.get() > 0 || successCount.get() != expectedExecutions || requestCount.get() != successCount.get()) {
+            fail("\nExpected Bulks: " + expectedExecutions + "\n" +
```
```suggestion
            if (exceptionRef.get() != null) {
                logger.error("exception(s) caught during test", exceptionRef.get());
            }
            fail("\nExpected Bulks: " + expectedExecutions + "\n" +
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+            }
+            listener.onResponse(bulkResponse);
+        };
+        BulkProcessor bulkProcessor = new BulkProcessor(consumer, BackoffPolicy.noBackoff(),
```
I think it is worth replacing the explicit close of the bulk processor below with a try-with-resources

```suggestion
        try (BulkProcessor bulkProcessor = new BulkProcessor(consumer, BackoffPolicy.noBackoff(),
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        final int concurrentBulkRequests = randomIntBetween(0, 20);
+        final int simulateWorkTimeInMillis = 5;
+        BulkResponse bulkResponse = new BulkResponse(new BulkItemResponse[]{new BulkItemResponse()}, 0);
+        AtomicInteger failureCount = new AtomicInteger(0);
```
```suggestion
        final AtomicReference<Exception> exceptionRef = new AtomicReference<>();
        AtomicInteger failureCount = new AtomicInteger(0);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+            }
+        };
+        ScheduledExecutorService flushExecutor = Executors.newScheduledThreadPool(1);
+        BulkProcessor bulkProcessor = new BulkProcessor(consumer, BackoffPolicy.noBackoff(),
```
Use a try-with-resources?

```suggestion
        try (BulkProcessor bulkProcessor = new BulkProcessor(consumer, BackoffPolicy.noBackoff(),
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                Scheduler.wrapAsScheduledCancellable(flushExecutor
+                    .schedule(command, delay.millis(), TimeUnit.MILLISECONDS)), flushExecutor::shutdownNow, BulkRequest::new);
+
+        ExecutorService executorService = Executors.newFixedThreadPool(concurrentClients);
```
```suggestion
        ExecutorService executorService = Executors.newFixedThreadPool(concurrentClients);
        CountDownLatch startGate = new CountDownLatch(1 + concurrentClients);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        IndexRequest indexRequest = new IndexRequest();
+        String bulkRequest = "{ \"index\" : { \"_index\" : \"test\", \"_id\" : \"1\" } }\n" +  "{ \"field1\" : \"value1\" }\n";
+        BytesReference bytesReference =
+            BytesReference.fromByteBuffers(new ByteBuffer[]{ByteBuffer.wrap(bulkRequest.getBytes(StandardCharsets.UTF_8))});
```
my personal preference so feel free to ignore

```suggestion
            BytesReference.fromByteBuffers(new ByteBuffer[] { ByteBuffer.wrap(bulkRequest.getBytes(StandardCharsets.UTF_8)) });
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                    //don't start any work until all tasks are submitted
+                    startGate.await();
+                    //alternate between ways to add to the bulk processor
+                    if (i.get() % 2 == 0) {
```
```suggestion
                    if (randomBoolean()) {
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                    }
+                }
+            }));
+        }
```
```suggestion
        }
        startGate.countDown();
        startGate.await();
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                f.get(10, TimeUnit.SECONDS);
+            } catch (Exception e){
+                failureCount.incrementAndGet();
+                logger.error("failure while getting future", e);
```
```suggestion
                logger.error("failure while getting future", e);
                exceptionRef.set(exceptionRef.get(), e);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        bulkProcessor.close();
+
+        if (failureCount.get() > 0 || requestCount.get() != successCount.get() || maxDocuments != docCount.get()) {
+            fail("\nRequested Bulks: " + requestCount.get() + "\n" +
```
```suggestion
            if (exceptionRef.get() != null) {
                logger.error("exception(s) caught during test", exceptionRef.get());
            }
            fail("\nRequested Bulks: " + requestCount.get() + "\n" +
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -118,4 +307,28 @@ public void afterBulk(long executionId, BulkRequest request, Throwable failure)
+            @Override
+            public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
+                if (failure != null) {
+                    failureCount.incrementAndGet();
```
We'd probably want to pass in the exceptionRef (and make it a throwable, but I thought we stopped catching throwable :|) and do the useOrSuppress thing here so we a view of all the exceptions during the test.

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+            TimeValue.timeValueMillis(simulateWorkTimeInMillis * 2),
+            (command, delay, executor) ->
+                Scheduler.wrapAsScheduledCancellable(flushExecutor
+                    .schedule(command, delay.millis(), TimeUnit.MILLISECONDS)), flushExecutor::shutdownNow, BulkRequest::new);
```
I wonder if we should wait for an ongoing flush to terminate

```suggestion
                    .schedule(command, delay.millis(), TimeUnit.MILLISECONDS)), () -> {
                         flushExecutor.shutdown();
                         try {
                             flushExecutor.awaitTermination(30L, TimeUnit.SECONDS);
                             if (flushExecutor.isTerminated() == false) {
                                 flushExecutor.shutdownNow();
                             }
                         } catch (InterruptedException e) {
                             Thread.currentThread().interrupt();
                         }
                    }, BulkRequest::new);
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+            try {
+                Thread.sleep(simulateWorkTimeInMillis); //simulate work
+            } catch (InterruptedException e) {
+                failureCount.getAndIncrement();
```
```suggestion
                Thread.currentThread().interrupt();
                failureCount.getAndIncrement();
```

jaymode commented on server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java:
```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+                Thread.sleep(simulateWorkTimeInMillis); //simulate work
+                listener.onResponse(bulkResponse);
+            } catch (InterruptedException e) {
+                failureCount.getAndIncrement();
```
```suggestion
                Thread.currentThread().interrupt();
                failureCount.getAndIncrement();
```

jaymode reviewed (changes requested):
I left some more comments and suggestions. Mainly concerning the handling of InterruptedException, keeping track of all exceptions, and tightening up the start gate latch so we know all items have been added and the executor has all of its threads constructed and has started running the expected number of concurrent tasks.

jakelandis committed:
Merge branch 'master' into bulk_processor_concurrent_requests

jakelandis committed:
better start gate, exception handling, interrupting cow, and formatting

jakelandis committed:
move onResponse inside try block

jakelandis commented:
@jaymode - thanks so much for the detailed review. Mind taking another look ?

I have addressed all the concerns with the tests (mostly on https://github.com/elastic/elasticsearch/pull/41451/commits/b446087bece587b00b7fb68a87cd4f07bed9123d)
* Better start gate to ensure all threads are running before test starts
* Capture all exceptions that may happen (reduced down to a single log event too)
* Properly propagate interrupted state
* Tweaked IDEA formatter and re-formatted. 
* Other minor improvements

jakelandis commented:
@elasticmachine run elasticsearch-ci/1

jaymode reviewed (approved):
LGTM. Thanks for the iterations @jakelandis

jakelandis commented:
@elasticmachine run elasticsearch-ci/1

bruckner commented:
Is there any chance this fix will be backported to 6.8?  Seems on par with the later #48697 which was backported.
